Introducing the pure go driver for sqlite

viscropst opened the pull request:
/close #556  
and related to k0sproject/k0s#6730.

brandond reviewed (changes requested):
Please add a test for this. You'll likely need to build a separate `-nocgo` image tag and/or binary, and then use that in the tests.

brandond commented on Dockerfile:
```diff
@@ -90,3 +122,16 @@ VOLUME /db
+VOLUME /db
+EXPOSE 2379/tcp
+USER nobody
+ENTRYPOINT ["/bin/kine"]
```
same, just add args to the existing stage

brandond commented on Dockerfile:
```diff
@@ -42,7 +43,10 @@ FROM scratch AS binary
+ARG CGO_ENABLED=1
+ARG BIN_SUFFIX
+ENV BIN_SUFFIX=${BIN_SUFFIX} CGO_ENABLED=${CGO_ENABLED}
+COPY --from=build /go/src/github.com/k3s-io/kine/bin/kine$BIN_SUFFIX /bin/kine
```
nit
```suggestion
COPY --from=build /go/src/github.com/k3s-io/kine/bin/kine${BIN_SUFFIX} /bin/kine
```

brandond commented on .github/workflows/pr.yml:
```diff
@@ -92,8 +145,14 @@ jobs:
     - name: Load Images
       run: |
         docker load --input ${{ runner.temp }}/kine.tar
+        docker load --input ${{ runner.temp }}/kine-nocgo.tar
```
Only load the nocgo image, and only run the test with the nocgo image.

brandond commented:
CI looks good, lint has some complaints about the code though:
```
#13 88.13 pkg/drivers/sqlite/sqlite.go:44:2: var-naming: var dsnUrl should be dsnURL (revive)
#13 88.13 	dsnUrl, _ := url.Parse(dsn)
#13 88.13 	^
#13 88.13 pkg/drivers/sqlite/sqlite.go:56:1: File is not properly formatted (gofmt)
#13 88.13 	query.Set("_txlock", "immediate") 
#13 88.13 ^
#13 88.13 pkg/drivers/sqlite/sqlite_test.go:14:1: File is not properly formatted (gofmt)
#13 88.13 	dsn,_ := getDataSourceName("")
#13 88.13 ^
```

brandond commented:
Sorry, I pushed another round of changes to CI; would you mind rebasing again? I think we're very close to done with this.

viscropst commented:
> Sorry, I pushed another round of changes to CI; would you mind rebasing again? I think we're very close to done with this.

Don't mind. It's normal for rapid CI migration from dedicated server to github CI.

brandond commented:
I am planning to merge the NATS fix PR https://github.com/k3s-io/kine/pull/549 before this one, so you will likely need to rebase one more time.

Would you like to add `-nocgo` variants to the release workflow, to produce pure-go images and binaries?

viscropst commented:
> I am planning to merge the NATS fix PR #549 before this one, so you will likely need to rebase one more time.
> 
> Would you like to add `-nocgo` variants to the release workflow, to produce pure-go images and binaries?

For add `-nocgo` variants to release workflow, and rebae is yes. I want to add the test target in Makefile for convenient local testing, Can I add them in Makefile ?

viscropst commented:
> Rebase looks good, can you squash to 2-3 commits as well?

I'm trying to do that.

brandond reviewed (commented):
one nit on what looks like an accidental action version downgrade - looks good other than that. Make sure to keep this last fix squashed and I think we can merge when it's addressed.

brandond commented:
Oh hmm, something seems to have broken sqlite for both cgo and nocgo. Did something get mangled during the squash? Or is it related to how you converted the query params to pragmas?

viscropst commented on .github/workflows/pr.yml:
```diff
@@ -98,7 +137,7 @@ jobs:
     steps:
     - name: Download Artifacts
-      uses: actions/download-artifact@v7
+      uses: actions/download-artifact@v4
```
It's my accident.

brandond commented:
seems sqlite tests are all still failing; I believe they were passing before you squashed. Please make sure nothing got mangled along the way.

viscropst commented on scripts/build:
```diff
@@ -22,4 +22,8 @@ LINKFLAGS="-X github.com/k3s-io/kine/pkg/version.Version=$VERSION"
-CGO_CFLAGS="-DSQLITE_ENABLE_DBSTAT_VTAB=1 -DSQLITE_USE_ALLOCA=1" go build -ldflags "$LINKFLAGS $OTHER_LINKFLAGS" -tags nats -o bin/kine
\ No newline at end of file
+if [ "$CGO_ENABLED" = "0" ]; then 
+    CGO_ENABLED=0 go build -ldflags "$LINKFLAGS" -tags nats -o bin/kine-nocgo
```
`-DSQLITE_ENABLE_DBSTAT_VTAB=1` was precomiled to pure-go sqlite, but `-DSQLITE_USE_ALLOCA=1` was not compiled to pure-go sqlite. For details see [generated code](https://raw.githubusercontent.com/modernc-org/sqlite/refs/heads/master/lib/sqlite_windows.go) and [generator code](https://gitlab.com/cznic/libsqlite3/-/blob/master/generator.go?ref_type=heads#L164)

brandond commented on pkg/drivers/sqlite/sqlite_nocgo.go:
```diff
@@ -4,30 +4,85 @@ package sqlite
-func New(_ context.Context, _ *sync.WaitGroup, _ *drivers.Config) (bool, server.Backend, error) {
-	return false, nil, errNoCgo
+func postCompactSQL() string {
+	return `PRAGMA wal_checkpoint(PASSIVE)`
```
Why would we want a passive checkpoint in the pure-go driver, but full in the cgo driver? The point here is to force a checkpoint and sync the WAL. That may not (and likely will not) happen with a passive checkpoint.

brandond replied:
@viscropst are you still interested in this?

brandond reviewed (changes requested):
one more question about the checkpoint mode.

I will probably merge this after new years when I will be around to tag a release and fix CI if anything in the release pipeline does not work when tested.

viscropst committed:
Introduce the pure go build to workflows, Dockerfiles and build scripts

Introduce the nocgo build to workflow pr

Introduce the nocgo to Dockerfile and build scripts

Fix the `multi-arch-package-nocgo` to use the binary from nocgo build

Introduce the nocgo packge for testing nocgo build

Use the "BIN_SUFFIX" to resolve the variant without cgo in build scripts

Unify the step names for `nocgo` variant

Unify the variable reference usages in `Dockerfile`

Split the sqlite test out of the others

Fix the syntax in pr workflow

Introduce the `nocgo` variant for `Dockerfile.test` building

Change back the job names to `docker` prefixed in pr workflow

Make `docker-nocgo` job load `nocgo` image only

Add back the sqlite test for `docker` job

Introduce `CGO_ENABLED` to `build-api-server-tests`

Introduce the `test` target to `Makefile` for local test

Introduce the `nocgo` variant to release workflow

Fix the missing `-nocgo` tag to `nocgo` variant

Change the built binary name to `kine${BIN_SUFFIX}-${ARCH}`

Signed-off-by: viscropst <viscropst@petalmail.com>

viscropst committed:
Introduce the pure go driver for sqlite

Split the default dsn for nocgo sqlite driver processing

Unify the `dataSourceName` to `getDataSourceName` for both cgo enabled and cgo disabled

Fix the naming and formatting reported by lint check for sqlite driver

Remove the unused `+build cgo` in `sqlite_cgo.go`

Unify the `_busy_timeout` processing to mattn's for easier modify the `busy_timeout`

Use the `PASSIVE` to checkpoint for nocgo variant to skip the busy invokers

Signed-off-by: viscropst <viscropst@petalmail.com>

viscropst committed:
Unify the `postCompact` function to all sqlite variant and introduce `_wal_autocheckpoint` parameter to change checkpoint mode

Signed-off-by: viscropst <viscropst@petalmail.com>

viscropst committed:
Change the `postCompact` mode to `FULL` for nocgo version

Signed-off-by: viscropst <viscropst@petalmail.com>

brandond commented:
ping @viscropst

brandond commented:
I'm going to close this, if you ever feel like picking up work on it again please feel free to reopen or start over with a new PR.

brandond commented:
reworked as:
* https://github.com/k3s-io/kine/pull/678
